gpbft/gpbft.go

```diff
@@ -1026,40 +1027,49 @@ type convergeState struct {
 	// Chains indexed by head CID
 	values map[TipSet]ECChain
 	// Tickets provided by proposers of each chain.
-	tickets map[TipSet][]Ticket
+	tickets map[TipSet]map[ActorID]Ticket
 }
 
 func newConvergeState() *convergeState {
 	return &convergeState{
 		values:  map[TipSet]ECChain{},
-		tickets: map[TipSet][]Ticket{},
+		tickets: map[TipSet]map[ActorID]Ticket{},
 	}
 }
 
 // Receives a new CONVERGE value from a sender.
-func (c *convergeState) Receive(value ECChain, ticket Ticket) error {
+func (c *convergeState) Receive(sender ActorID, value ECChain, ticket Ticket) error {
 	if value.IsZero() {
 		return fmt.Errorf("bottom cannot be justified for CONVERGE")
 	}
 	key := value.Head()
 	c.values[key] = value
-	c.tickets[key] = append(c.tickets[key], ticket)
+
+	if _, ok := c.tickets[key]; !ok {
+		c.tickets[key] = make(map[ActorID]Ticket)
+	}
+
+	c.tickets[key][sender] = ticket
 
 	return nil
 }
 
-func (c *convergeState) findMinTicketProposal() ECChain {
-	var minTicket Ticket
-	var minValue ECChain
+func (c *convergeState) findMaxTicketProposal(table PowerTable) ECChain {
+	var maxTicket *big.Int
+	var maxValue ECChain
+
 	for cid, value := range c.values {
```

Comment:
I have realised that if the same ticket is used for two different values (equivocating), then the one selected here will be nondeterministic, governed by Go map iteration order. We should break ties deterministically instead. We can use the ECChain value itself to do that if we make them comparable. (You can file an issue for this instead of fixing it here if you want).

Reply:
I think it is ok to have non-determinism here. If the same ticket is used for two different values then either we get a decision on one of them only or we go to a new round. Eventually there is a round where the max ticket is held by a correct, which will not double vote.

Reply:
Ok, thanks. As we discussed in person, this is ok but just needs to be clearly documented in the spec (and ideally here too) so readers understand that.